### ISSUE TITLE
api_dump: Consider not nesting pNext chains in human-readable output

### ISSUE THREAD

**Issue by karl-lunarg** [LunarG/VulkanTools#973]
If api_dump encounters an API call with a very long pNext chain, api_dump prints it out like the example at the bottom of this comment.  There are two problems with this:

* The horizontal indenting makes it pretty hard to read.
* The interesting data members in the same struct as their corresponding `sType` and `pNext` fields are printed out very far apart from each other.  It takes a lot of effort to "match up" the data members with their structure type in order to interpret them.

It might be better to have it look something like:
```
sType:         VkStructureType = blah
pNext:         blah
member1:       123
member2:       xyz
sType:         VkStructureType = another blah
pNext:         another blah
member1:       123
member2:       xyz
```

It can also be argued that these pNext chains really are not nested structures.  They are a pointer-chained list of structures and shouldn't be displayed in a manner that suggests nesting, as is done today.

I understand that there may be some elegant common code generation steps that happen to do things this way, but in this case, it may be better to treat this as a special case.

This is a pretty low priority and somewhat subjective suggestion.  So if there is some other good reason for the current behavior, please disregard it.

```
vkGetPhysicalDeviceFeatures2(physicalDevice, pFeatures) returns void:
    physicalDevice:                 VkPhysicalDevice = 0x558c32a33660
    pFeatures:                      VkPhysicalDeviceFeatures2* = 0x558c32a32640:
        sType:                          VkStructureType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_FEATURES_2 (1000059000)
        pNext:                          VkPhysicalDeviceSamplerYcbcrConversionFeatures = 0x558c32a32730:
            sType:                          VkStructureType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_SAMPLER_YCBCR_CONVERSION_FEATURES (1000156004)
            pNext:                          VkPhysicalDevice8BitStorageFeatures = 0x558c32a32748:
                sType:                          VkStructureType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_8BIT_STORAGE_FEATURES (1000177000)
                pNext:                          VkPhysicalDevice16BitStorageFeatures = 0x558c32a32768:
                    sType:                          VkStructureType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_16BIT_STORAGE_FEATURES (1000083000)
                    pNext:                          VkPhysicalDeviceVariablePointersFeatures = 0x558c32a32788:
                        sType:                          VkStructureType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_VARIABLE_POINTERS_FEATURES (1000120000)
                        pNext:                          VkPhysicalDeviceDescriptorIndexingFeatures = 0x558c32a327b8:
                            sType:                          VkStructureType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_DESCRIPTOR_INDEXING_FEATURES (1000161001)
                            pNext:                          VkPhysicalDeviceInlineUniformBlockFeaturesEXT = 0x558c32a32818:
                                sType:                          VkStructureType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_INLINE_UNIFORM_BLOCK_FEATURES_EXT (1000138000)
                                pNext:                          VkPhysicalDeviceShaderFloat16Int8Features = 0x558c32a32898:
                                    sType:                          VkStructureType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_SHADER_FLOAT16_INT8_FEATURES (1000082000)
                                    pNext:                          VkPhysicalDeviceVertexAttributeDivisorFeaturesEXT = 0x558c32a327a0:
                                        sType:                          VkStructureType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_VERTEX_ATTRIBUTE_DIVISOR_FEATURES_EXT (1000190002)
                                        pNext:                          VkPhysicalDeviceVulkanMemoryModelFeatures = 0x558c32a32830:
                                            sType:                          VkStructureType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_VULKAN_MEMORY_MODEL_FEATURES (1000211000)
                                            pNext:                          VkPhysicalDeviceShaderAtomicInt64Features = 0x558c32a32850:
                                                sType:                          VkStructureType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_SHADER_ATOMIC_INT64_FEATURES (1000180000)
                                                pNext:                          VkPhysicalDeviceConditionalRenderingFeaturesEXT = 0x558c32a32868:
                                                    sType:                          VkStructureType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_CONDITIONAL_RENDERING_FEATURES_EXT (1000081001)
                                                    pNext:                          VkPhysicalDeviceScalarBlockLayoutFeatures = 0x558c32a32880:
                                                        sType:                          VkStructureType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_SCALAR_BLOCK_LAYOUT_FEATURES (1000221000)
                                                        pNext:                          void* = NULL
                                                        scalarBlockLayout:              VkBool32 = 1
                                                    conditionalRendering:           VkBool32 = 1
                                                    inheritedConditionalRendering:  VkBool32 = 1
                                                shaderBufferInt64Atomics:       VkBool32 = 1
                                                shaderSharedInt64Atomics:       VkBool32 = 1
                                            vulkanMemoryModel:              VkBool32 = 1
                                            vulkanMemoryModelDeviceScope:   VkBool32 = 1
                                            vulkanMemoryModelAvailabilityVisibilityChains: VkBool32 = 1
                                        vertexAttributeInstanceRateDivisor: VkBool32 = 1
                                        vertexAttributeInstanceRateZeroDivisor: VkBool32 = 1
                                    shaderFloat16:                  VkBool32 = 0
                                    shaderInt8:                     VkBool32 = 1
                                inlineUniformBlock:             VkBool32 = 1
                                descriptorBindingInlineUniformBlockUpdateAfterBind: VkBool32 = 1
                            shaderInputAttachmentArrayDynamicIndexing: VkBool32 = 1
                            shaderUniformTexelBufferArrayDynamicIndexing: VkBool32 = 1
                            shaderStorageTexelBufferArrayDynamicIndexing: VkBool32 = 1
                            shaderUniformBufferArrayNonUniformIndexing: VkBool32 = 1
                            shaderSampledImageArrayNonUniformIndexing: VkBool32 = 1
                            shaderStorageBufferArrayNonUniformIndexing: VkBool32 = 1
                            shaderStorageImageArrayNonUniformIndexing: VkBool32 = 1
                            shaderInputAttachmentArrayNonUniformIndexing: VkBool32 = 1
                            shaderUniformTexelBufferArrayNonUniformIndexing: VkBool32 = 1
                            shaderStorageTexelBufferArrayNonUniformIndexing: VkBool32 = 1
                            descriptorBindingUniformBufferUpdateAfterBind: VkBool32 = 0
                            descriptorBindingSampledImageUpdateAfterBind: VkBool32 = 1
                            descriptorBindingStorageImageUpdateAfterBind: VkBool32 = 1
                            descriptorBindingStorageBufferUpdateAfterBind: VkBool32 = 1
                            descriptorBindingUniformTexelBufferUpdateAfterBind: VkBool32 = 1
                            descriptorBindingStorageTexelBufferUpdateAfterBind: VkBool32 = 1
                            descriptorBindingUpdateUnusedWhilePending: VkBool32 = 1
                            descriptorBindingPartiallyBound: VkBool32 = 1
                            descriptorBindingVariableDescriptorCount: VkBool32 = 1
                            runtimeDescriptorArray:         VkBool32 = 1
                        variablePointersStorageBuffer:  VkBool32 = 1
                        variablePointers:               VkBool32 = 1
                    storageBuffer16BitAccess:       VkBool32 = 1
                    uniformAndStorageBuffer16BitAccess: VkBool32 = 1
                    storagePushConstant16:          VkBool32 = 1
                    storageInputOutput16:           VkBool32 = 0
                storageBuffer8BitAccess:        VkBool32 = 1
                uniformAndStorageBuffer8BitAccess: VkBool32 = 1
                storagePushConstant8:           VkBool32 = 1
            samplerYcbcrConversion:         VkBool32 = 1
        features:                       VkPhysicalDeviceFeatures = 0x558c32a32650:
            robustBufferAccess:             VkBool32 = 1
...
```

**Comment by charles-lunarg** [comment]
The reason the code looks like this is because the pNext code works by piggingbacking off the existing struct dumping functions. That made the code very simple to write and didn't cause special cases, as a "pNext member" simply calls recursively into the chain. 

I agree, the formatting leaves a lot to be desired, with the struct type being seperated from the contents and a long pNext chain making it very hard to read. The solution would be to make struct printing special case when inside a pNext chain, but that will involve making the struct dumping function need to know, which introduces more complexity.

**Comment by charles-lunarg** [comment]
PR #1114 Should address this for text output, but doesn't change HTML or JSON. The reason is that HTML/JSON are structured so that structs being printed will always have a 'begin and 'end' sequence. The way pNext Chains are interpreted is as a 'subobject', which means they will, without significant modifications to the code, always be printed in a nested format. The text output differs in that it lacks the begin/end, allowing for a sidechannel that simply doesn't indent the pNext chain as it prints. In short, Text is more amenable to making it not indent for every level, while HTML/JSON isn't.

**Comment by karl-lunarg** [comment]
Thanks for doing this.  Can you post a sample of what the new output looks like, either here or in the PR?

Also, I think doing this only for the text output makes sense from a user's point of view in addition to the reasoning you provided.

**Comment by charles-lunarg** [comment]
Formatting is very basic, but makes it a straightforward list of structs.
```
Thread 0, Frame 0, Time 9168 us:
vkGetPhysicalDeviceSurfaceCapabilities2KHR(physicalDevice, pSurfaceInfo, pSurfaceCapabilities) returns VkResult VK_SUCCESS (0):
    physicalDevice:                 VkPhysicalDevice = 0x55859af34c40
    pSurfaceInfo:                   const VkPhysicalDeviceSurfaceInfo2KHR* = 0x7ffc12999c10:
        sType:                          VkStructureType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_SURFACE_INFO_2_KHR (1000119000)
        pNext:                          const void* = NULL
        surface:                        VkSurfaceKHR = 0x55859af3d910
    pSurfaceCapabilities:           VkSurfaceCapabilities2KHR* = 0x55859af48d48:
        sType:                          VkStructureType = VK_STRUCTURE_TYPE_SURFACE_CAPABILITIES_2_KHR (1000119001)
        pNext:                          const void* = VkSharedPresentSurfaceCapabilitiesKHR
        surfaceCapabilities:            VkSurfaceCapabilitiesKHR = 0x55859af48d58:
            minImageCount:                  uint32_t = 2
            maxImageCount:                  uint32_t = 8
            currentExtent:                  VkExtent2D = 0x55859af48d60:
                width:                          uint32_t = 256
                height:                         uint32_t = 256
            minImageExtent:                 VkExtent2D = 0x55859af48d68:
                width:                          uint32_t = 256
                height:                         uint32_t = 256
            maxImageExtent:                 VkExtent2D = 0x55859af48d70:
                width:                          uint32_t = 256
                height:                         uint32_t = 256
            maxImageArrayLayers:            uint32_t = 1
            supportedTransforms:            VkSurfaceTransformFlagsKHR = 1 (VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR)
            currentTransform:               VkSurfaceTransformFlagBitsKHR = 1 (VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR)
            supportedCompositeAlpha:        VkCompositeAlphaFlagsKHR = 1 (VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR)
            supportedUsageFlags:            VkImageUsageFlags = 159 (VK_IMAGE_USAGE_TRANSFER_SRC_BIT | VK_IMAGE_USAGE_TRANSFER_DST_BIT | VK_IMAGE_USAGE_SAMPLED_BIT | VK_IMAGE_USAGE_STORAGE_BIT | VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT | VK_IMAGE_USAGE_INPUT_ATTACHMENT_BIT)
        pNext:                          VkSharedPresentSurfaceCapabilitiesKHR = 0x55859af53d90:
            sType:                          VkStructureType = VK_STRUCTURE_TYPE_SHARED_PRESENT_SURFACE_CAPABILITIES_KHR (1000111000)
            pNext:                          const void* = VkSurfaceProtectedCapabilitiesKHR
            sharedPresentSupportedUsageFlags: VkImageUsageFlags = 0
        pNext:                          VkSurfaceProtectedCapabilitiesKHR = 0x55859af53db0:
            sType:                          VkStructureType = VK_STRUCTURE_TYPE_SURFACE_PROTECTED_CAPABILITIES_KHR (1000239000)
            pNext:                          const void* = NULL
            supportsProtected:              VkBool32 = 0
```

**Comment by charles-lunarg** [comment]
Merged #1114 
Closing the issue.
